pkg/kubelet/prober/runner.go

```diff
@@ -0,0 +1,255 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prober
+
+import (
+	"context"
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/probe"
+	httpprobe "k8s.io/kubernetes/pkg/probe/http"
+	"net/http"
+	"time"
+)
+
+// probeRunner is an interface for running probes, objects implementing
+// probeRunner encapsulates the logic of running probes.
+type probeRunner interface {
```

Comment:
From an implementation perspective if the goal is to reuse objects, why not use a sync.Pool much closer to the point where a request is created?  That prevents this higher layer from being visible to other components (ie a prober is responsible for optimizing its own memory use).

Reply:
If we use a single shared pool at prober, we would need to mutate the request object to meet the containers probe spec(IP, port, path) for every probe and every container which might consume some CPU.

Reply:
this ^^^. Maybe we can hide implementation of this new object that holds the request object into the proper packages.